feat: support CLI multiple projects (#2086)
wix/stylable: packages/cli/README.md

`*` - For the `useNamespaceReference` flag to function properly, the `source` folder must be published in addition to the output `target` code

### Configuration file

> `stc` configuration file should be defined in the `stylable.config.js` under the name `stcConfig`.

Reviewer: ```suggestion
The `stc` configuration should be located in the `stylable.config.js` file under the property name `stcConfig`.
```

> We also provide a helper method and types to give you a better configuration experience.

Reviewer: ```suggestion
The CLI provides a helper method and type definitions to provide a better configuration experience.
```

$ stc --srcDir="./src" --outDir="./dist"
```

## Multiple Projects

> Projects allow you to share `stc` configuration and manage your Stylable projects in one place with predictable and controlled build order with caching optimizations.

Reviewer: ```suggestion
Projects allow sharing `stc` configurations and management of Stylable projects in one location. They provides a controllable and predictable build order with caching optimizations.
```

```ts
export interface MultipleProjectsConfig<PRESET extends string> {
    options?: PartialBuildOptions;
    presets?: Presets<PRESET>;
    projects: Projects<PRESET>;
    projectsOptions?: {
        resolveRequests?: ResolveRequests;
    };
}
```

> Example for simple monorepo with Stylable packages
```js
const { typedConfiguration } = require('@stylable/cli');

exports.stcConfig = typedConfiguration({
    options: {
        srcDir: './src',
        outDir: './dist',
        outputSources: true,
        cjs: false,
        useNamespaceReference: true, 
    },
    projects: ['packages/*']
});

```

### Options

> Like for [single project](#configuration-file) `options` is the top-level `BuildOptions` and is the default options for each project.

Reviewer: ```suggestion
Similar to a [single project](#configuration-file), `options` is the top-level `BuildOptions` and is the default options for each project.
```

### Projects

> **Projects** is a generic term that refers to the set of requests with defined single or multiple `BuildOptions`.\

Reviewer: ```suggestion
**Projects** is a generic term that refers to a set of path requests that define single or multiple `BuildOptions`.\
```

This set of requests is being processed and then evaluated as a map of `projectRoot` (directory path) to a set of `BuildOptions`.

By default, the request is a path to a package, and in order to make the correct topological sort, the dependency needs to be specified in each package `package.json`

> As said, the value of the request can be resolved to a single `BuildOptions`, but it can also be multiple builds.

Reviewer: ```suggestion
As mentioned above, the value of a request can be resolved to a single or multiple `BuildOptions`.
```

```jsonc
{
    //...
    projects: {
        "packages/*": { 
            // ...BuildOptions
        },
        "other-package/*": [
            { /* #1 ...BuildOptions */ }, 
            { /* #2 ...BuildOptions */ }, 
        ]
    }
    //...
}
```

> The full types specification for defining Projects
```ts
export type Projects =
    | Array<string | [string, ProjectEntryValues]>
    | Record<string, ProjectEntryValues>;

export type ProjectEntryValues<PRESET extends string> =
    | ProjectEntryValue<PRESET>
    | Array<ProjectEntryValue<PRESET>>;

export type ProjectEntryValue<PRESET extends string> =
    | PRESET
    | PartialBuildOptions
    | {
          preset?: PRESET;
          presets?: Array<PRESET>;
          options: PartialBuildOptions;
      };
```

### Presets

> To reuse `BuildOptions`, you can define them under a name inside the `presets` property and use them inside the project entry value.

Reviewer: ```suggestion
To reuse `BuildOptions`, define them using a name under the `presets` property and use them as the project entry value.
```

```js
exports.stcConfig = {
    //...
    presets: {
        firstPreset: {/* ...BuildOptions */},
        secondPreset: {/* ...BuildOptions */},
    },
    projects: {
        'packages/*': ['firstPreset', 'secondPreset']
    }

};
```

### Projects Options

> These options let you control the projects resolution process.

Reviewer: ```suggestion
These options control the projects resolution process.
```

    },
    projects: ['packages/*']
});

```

### Options

Like for [single project](#configuration-file) `options` is the top-level `BuildOptions` and is the default options for each project.

### Projects

**Projects** is a generic term that refers to the set of requests with defined single or multiple `BuildOptions`.\
This set of requests is being processed and then evaluated as a map of `projectRoot` (directory path) to a set of `BuildOptions`.

By default, the request is a path to a package, and in order to make the correct topological sort, the dependency needs to be specified in each package `package.json`

As said, the value of the request can be resolved to a single `BuildOptions`, but it can also be multiple builds.

```jsonc
{
    //...
    projects: {
        "packages/*": { 
            // ...BuildOptions
        },
        "other-package/*": [
            { /* #1 ...BuildOptions */ }, 
            { /* #2 ...BuildOptions */ }, 
        ]
    }
    //...
}
```

> The full types specification for defining Projects
```ts
export type Projects =
    | Array<string | [string, ProjectEntryValues]>
    | Record<string, ProjectEntryValues>;

// P = Preset name

> export type ProjectEntryValue<P extends string> =

Reviewer 2: ```suggestion
export type ProjectEntryValue<PRESET extends string> =
```

    | P
    | PartialBuildOptions
    | {
          preset?: P;
          presets?: Array<P>;
          options: PartialBuildOptions;
      };

> export type ProjectEntryValues<P extends string> =

Reviewer 2: maybe switch `ProjectEntryValues` to be above `ProjectEntryValue` as it is used first and then uses the other.
